src/lib.rs

```diff
@@ -143,12 +156,49 @@ impl<'a> EbpfVmMbuff<'a> {
     /// ];
     ///
     /// // Instantiate a VM.
-    /// let mut vm = rbpf::EbpfVmMbuff::new(prog1);
+    /// let mut vm = rbpf::EbpfVmMbuff::new(Some(prog1));
     /// vm.set_prog(prog2);
     /// ```
     pub fn set_prog(&mut self, prog: &'a [u8]) {
-        verifier::check(prog);
-        self.prog = prog;
+        (self.verifier)(prog);
+        self.prog = Some(prog);
+    }
+
+    /// Sets a new verifier function
+    ///
+    /// # Panics
+    ///
+    /// The simple verifier may panic if it finds errors in the eBPF program at load time.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use rbpf::ebpf;
+    ///
+    /// // simple verifier
+    /// fn verifier(prog: &[u8]) -> bool {
+    ///     let last_insn = ebpf::get_insn(prog, (prog.len() / ebpf::INSN_SIZE) - 1);
+    ///     if last_insn.opc != ebpf::EXIT {
+    ///         panic!("[Verifier] Error: program does not end with “EXIT” instruction");
+    ///     }
+    ///     true
+    /// }
+    ///
+    /// let prog1 = &[
+    ///     0xb7, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, // mov r0, 0
+    ///     0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00  // exit
+    /// ];
+    ///
+    /// // Instantiate a VM.
+    /// let mut vm = rbpf::EbpfVmMbuff::new(Some(prog1));
+    /// // Change the verifier
+    /// vm.set_verifier(verifier);
+    /// ```
+    pub fn set_verifier(&mut self, verifier: Verifier) {
+        if let Some(prog) = self.prog {
+            verifier(prog);
+        }
```

Comment:
So you would have the verifier run when we set it, if a program has already been set. Hmm that's interesting. My first opinion was that it does not make sense to load a program before setting a verifier, because in that case you first check the program with the default verifier that has nothing to do with the custom function you want to load. So I was nearly tempted to reject setting the verifier if a program had been loaded already.

On the other hand, with your code we could now imagine having several _complementary_ verifiers that we would like to run in turns on a program... Turns out I like this idea. Let's keep this :)